libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/controller/StateResetController.kt

```diff
@@ -0,0 +1,37 @@
+package com.mapbox.navigation.ui.app.internal.controller
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.internal.extensions.flowTripSessionState
+import com.mapbox.navigation.core.trip.session.TripSessionState
+import com.mapbox.navigation.ui.app.internal.Store
+import com.mapbox.navigation.ui.base.lifecycle.UIComponent
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.debounce
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class, FlowPreview::class)
+internal class StateResetController(private val store: Store) : UIComponent() {
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        // TripSessionState might change in quick succession.
+        // Debouncing it by 100 ms ensures we skip transient values.
+        var prevState = mapboxNavigation.getTripSessionState()
+        mapboxNavigation.flowTripSessionState().debounce(100).observe { newState ->
```

Comment:
The main concern with `debounce` is what if we have a slow device and more than 100ms passes when you change the replay setting. We will reset `SharedApp` state, but will we go back to the previous state after a new trip session is started?

Reply:
I am still curious

```
mapboxNavigation.flowTripSessionState().debounce(100).observe { newState ->
    if (prevState == TripSessionState.STARTED && newState == TripSessionState.STOPPED) {
        // we only reset Store state when TripSessionState switches from STARTED to STOPPED.
        // mapboxNavigation.setNavigationRoutes(emptyList())
        store.reset()
    }
    prevState = newState
}
```
`store.reset()` resets the `State()` such that `State.routes` is now `RoutesState.Empty`. In this case what is then setting the route back? Probably it's because routes on mapboxNavigation is still available

Reply:
@abhishek1508 yup. That's exactly what happens. When a user clicks the "END NAVIGATION" action in the foreground notification, MapboxNavigation stops the trip session (and all background services). but it does not remove the assigned route, hence the need for `mapboxNavigation.setNavigationRoutes(emptyList())` call.

Reply:
@Zayankovsky I agree with you. I do not like the use of `debounce` here. It does feel hacky. However, it allows us to skip the transient TripSession state when toggling between: 

```
MapboxNavigation.startTripSession() -> MapboxNavigation.stopTripSession() -> MapboxNavigation.startReplayTripSession()
```

and vice versa.

Reply:
Resolving the thread, as I don't see how this can be implemented in a better way at the moment.